hc_dpki/src/key_blob.rs

```diff
@@ -0,0 +1,178 @@
+    ///  base64 encoded, json serialized string of the EncryptedData
+    pub data: String,
+}
+
+impl KeyBundle {
+    /// Generate an encrypted blob for persistence
+    /// @param {SecBuf} passphrase - the encryption passphrase
+    /// @param {string} hint - additional info / description for the bundle
+    /// @param {string} config - Settings for pwhash
+    pub fn as_blob(
+        &mut self,
+        passphrase: &mut SecBuf,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> HcResult<KeyBlob> {
+        // Initialize buffer
+        let mut data_buf = SecBuf::with_secure(BLOB_DATA_LEN);
+        let mut offset: usize = 0;
+        // Write version
+        data_buf.write(offset, &[BLOB_FORMAT_VERSION])?;
+        offset += 1;
+        // Write public signing key
+        let key = self.sign_keys.decode_pub_key();
+        assert_eq!(sign::PUBLICKEYBYTES, key.len());
+        data_buf
+            .write(offset, &key)
+            .expect("Failed blobbing public signing key");
+        offset += sign::PUBLICKEYBYTES;
+        // Write public encoding key
+        let key = self.enc_keys.decode_pub_key();
+        assert_eq!(kx::PUBLICKEYBYTES, key.len());
+        data_buf
+            .write(offset, &key)
+            .expect("Failed blobbing public encoding key");
+        offset += kx::PUBLICKEYBYTES;
+        // Write private signing key
+        data_buf
+            .write(offset, &**self.sign_keys.private.read_lock())
+            .expect("Failed blobbing private signing key");
+        offset += sign::SECRETKEYBYTES;
+        // Write private encoding key
+        data_buf
+            .write(offset, &**self.enc_keys.private.read_lock())
+            .expect("Failed blobbing private encoding key");
+        offset += kx::SECRETKEYBYTES;
+        assert_eq!(offset, BLOB_DATA_LEN_MISALIGN);
+
+        // encrypt buffer
+        let encrypted_blob = password_encryption::pw_enc(&mut data_buf, passphrase, config)?;
+        let serialized_blob = json::encode(&encrypted_blob).expect("Failed to serialize KeyBundle");
+        // conver to base64
+        let encoded_blob = base64::encode(&serialized_blob);
+        // Done
+        Ok(KeyBlob {
+            seed_type: self.seed_type.clone(),
+            hint,
+            data: encoded_blob,
+        })
+    }
+
+    /// Construct the pairs from an encrypted blob
+    /// @param {object} bundle - persistence info
+    /// @param {SecBuf} passphrase - decryption passphrase
+    /// @param {string} config - Settings for pwhash
+    pub fn from_blob(
+        blob: &KeyBlob,
+        passphrase: &mut SecBuf,
+        config: Option<PwHashConfig>,
+    ) -> HcResult<KeyBundle> {
+        // decoding the blob.data of type EncryptedData
+        let blob_decoded = base64::decode(&blob.data)?;
+
+        // Deserialize
+        let blob_string = str::from_utf8(&blob_decoded).unwrap();
+        let data: EncryptedData = json::decode(&blob_string).unwrap();
```

Comment:
lets get this deserialize logic behind a `?` in case bad input

Reply:
Fixed and changed to serde_json